### ISSUE TITLE
Allow a per-particle setting of the dt time step parameter

### ISSUE THREAD

**Issue by erikvansebille** [OceanPARCELS/parcels#23]
For RK45 methods, it will be important to allow particles within a pset to have different dt time steps. It would be good to be able to implement this

**Comment by mlange05** [comment]
Ok, the key change here is that the current `time` and `dt` variable are stored/computed per particle, instead of being passed in through the outer kernel loops. This entails that the kernel argument signature changes, which the current JIT engine does not yet support. So, I believe there are two things to consider:
* In "SciPy mode" the outer loops are really trivial and we don't actually need to go through the `pset.execute()` call. This means anyone developing a new kernel with a different argument signature can write their own outer loops over time and particles and still use all the grid evaluation methods. So this call will be replaced with something like:
```
for _ in range(timesteps):
    for p in pset.particles:
        AdvectionRK45(p, pset.grid)  
```
* In JIT mode we need to consider a more dynamic way to derive kernel argument and insert algorithm specific variable, such as `current_time`, on the particles. To do this we probably want to create a new particle class `RK45Particle`, so that custom particles can inherit JIT and RK45 at the same time, ie. `class MyCustomParticle(JITParticle, RK45Particle):`. The current ParticleType class must then carry information about the kernel signature as well as the composition of the struct, and might adjust the given user code when generating a `Kernel` object from a given Python function. It might take a little time to get this to work, but it is definitely feasible. It also becomes much easier if we work from an existing "pure Python" version, since we know what the new kernel function looks like.

**Comment by erikvansebille** [comment]
OK, sounds like a plan to me. Let's first build an RK45 in pure python and then see how we can JIT that. Joel, are you up for that?

**Comment by Abobie** [comment]
Sure, sounds good

**Comment by Abobie** [comment]
Just a reminder that I need JIT to support `if/else` for RK45 to work (and `break` if that's not included yet).

**Comment by mlange05** [comment]
@Abobie PR #42 should provide most of the language features that you require for the current version of your RK45 kernels. Can you please review and test this PR?

However, we do have to think a bit further about the loop structures though. I don't think we want the time loop internalised in the kernel, since this means we cannot concatenate it with any other kernel. This then means that the signature of the kernels in the time loop needs to change so that the RK45 kernel can control the time stepping.

**Comment by Abobie** [comment]
@mlange05 I've been playing around with the additions in PR 
[#42](https://github.com/OceanPARCELS/parcels/pull/42), it seems to be working just fine! Regarding the loops, are you saying that I should try to move out the `while True` from `AdvectionRK45`? Or the `while` loop from `execute_adaptive` in `kernel.py`?

**Comment by mlange05** [comment]
@Abobie Ok, I see you now actually have the same function signature than the other kernels, so you might not actually need to remove the while loop from the kernel. By keeping it in I believe you are switching the outer-loop structure though with the particle loop being the outer one, which requires some changes to the JIT code path. We should discuss whether we want to do this for all kernels/particles and/or whether this becomes an optional feature that gets triggered when a user adds `AdvectionRK45` to a kernel chain.

**Comment by Abobie** [comment]
Ok, I see. I didn't think much about that when I wrote the code, this just seemed the most natural way to structure the loops for RK45.

**Comment by erikvansebille** [comment]
Implemented in #78
